syntax = "proto3";

import "google/protobuf/struct.proto";

package grpc_proto.optimizePortfolio;

option java_multiple_files = true;
option java_package = "com.xstock.grpcProto.optimizePortfolio";
option java_outer_classname = "OptimizedPortfolio";
option objc_class_prefix = "HLW";

service OptimizePortfolio {
  rpc Optimize (OptimizedPortfolioRequest) returns (OptimizedPortfolioResponse) {};
}

message Asset {
  string ticker = 1;
  repeated google.protobuf.Struct historical = 2;
}

message OptimizedPortfolioRequest {
  repeated Asset assets = 1;
  string objective = 2;
  uint64 capital = 3;
}

message WeightedAsset {
  string ticker = 1;
  optional float weight = 2;
  optional uint32 shares = 3;
  optional uint64 capital = 4;
}

message OptimizedPortfolioResponse {
  repeated WeightedAsset portfolio = 1;
  optional float expected_returns = 2;
  optional float risk = 3;
  uint64 remaining_capital = 4;
}